src/redisearch_rs/buffer/src/lib.rs

```diff
@@ -0,0 +1,314 @@
+/*
+ * Copyright (c) 2006-Present, Redis Ltd.
+ * All rights reserved.
+ *
+ * Licensed under your choice of the Redis Source Available License 2.0
+ * (RSALv2); or (b) the Server Side Public License v1 (SSPLv1); or (c) the
+ * GNU Affero General Public License v3 (AGPLv3).
+*/
+
+//! The `Buffer` API Rust wrappers.
+//!
+//! This crate provides Rust wrappers for the `Buffer`, `BufferReader` and `BufferWriter` structs
+//! from `buffer.h`.
+
+use std::{
+    cmp,
+    ptr::{NonNull, copy_nonoverlapping},
+    slice,
+};
+
+/// Redefines the `Buffer` struct from `buffer.h`
+///
+/// # Safety
+///
+/// * `data` must point to a valid, properly initialized `Buffer`.
+/// * `capacity` must not exceed the length of the Buffer.
+/// * `len` must not exceed the length of the Buffer.
```

Comment:
```suggestion
/// * `data` must point to a valid, properly initialized `Buffer`.
/// * `capacity` must not exceed the capacity of the Buffer.
/// * `len` must not exceed the length or capacity of the Buffer.
```

Reply:
The "length" here is the actual length/size of the buffer, so I think my docs are correct. :)

Reply:
Isn't length the size used of the available capacity? Hence the `debug_assert!` later to make sure `len <= cap`. This means that capacity should always equal or exceed the length of the buffer.

Reply:
> /// * `len` must not exceed the length or capacity of the Buffer

Hmm.. I think you've a point about this part though. I think this should be "`len` must not exceed the capacity"

Reply:
I update the docs here and also moved the safety comment for each field to their own docs, since the fields are no longer public.